python/lsst/dax/ppdb/bigquery/ppdb_bigquery.py

```diff
@@ -0,0 +1,417 @@
+                ApdbTables.DiaSource.value: sources,
+                ApdbTables.DiaForcedSource.value: forced_sources,
+            }
+
+            # Loop over the table data and write each table to a Parquet file.
+            for table_name, table_data in table_dict.items():
+                if not table_data.rows():
+                    _LOG.debug("No data for %s in %s, skipping export", table_name, replica_chunk.id)
+                    continue
+                parquet_file_path = chunk_dir / f"{table_name}.parquet"
+                try:
+                    with Timer(
+                        "write_parquet_time", _MON, tags={"table": table_name, "path": str(parquet_file_path)}
+                    ) as timer:
+                        row_count = write_parquet(
+                            table_name,
+                            table_data,
+                            parquet_file_path,
+                            batch_size=self.batch_size,
+                            compression_format=self.compression_format,
+                            exclude_columns={"apdb_replica_subchunk"},
+                        )
+                        timer.add_values(row_count=row_count)
+                    _LOG.info("Wrote %s with %d rows to %s", table_name, row_count, parquet_file_path)
+                except Exception:
+                    _LOG.exception("Failed to write %s", table_name)
+                    raise
+
+            # Create manifest for the replica chunk.
+            try:
+                manifest = self._generate_manifest(replica_chunk, table_dict)
+                _LOG.info("Generated manifest for %s: %s", replica_chunk.id, manifest.model_dump_json())
+            except Exception:
+                _LOG.exception("Failed to generate manifest for %d", replica_chunk.id)
+                raise
+
+            # Write manifest data to a JSON file.
+            try:
+                manifest.write_json_file(chunk_dir)
+            except Exception:
+                _LOG.exception("Failed to write manifest file for %d to %s", replica_chunk.id, chunk_dir)
+                raise
+        except Exception:
+            _LOG.exception("Failed to store replica chunk: %s", replica_chunk.id)
+            raise
+
+        # Store the replica chunk info in the database, including status and
+        # directory.
+        replica_chunk_ext = PpdbReplicaChunkExtended.from_replica_chunk(
+            replica_chunk, ChunkStatus.EXPORTED, chunk_dir
+        )
+        try:
+            self.store_chunk(replica_chunk_ext, False)
+        except Exception as e:
+            _LOG.exception("Failed to store replica chunk info in database for %s", replica_chunk.id)
+            raise e
+
+        _LOG.info("Done processing %s", replica_chunk.id)
+
+    def _get_chunk_path(self, chunk_id: int) -> Path:
+        path = Path(
+            self.directory,
+            datetime.datetime.today().strftime("%Y/%m/%d"),
```

Comment:
I think it would be better to use `ReplicaChunk.last_update_time` instead of `today` for generating path, that would avoid potential duplicates on retries.

Reply:
This was from a prior PR. Would it make sense to just have the local directory be the replica chunk ID, since that is supposed to be unique?

Retries are not really well-handled yet either here or in the cloud functions; if the system encounters the same replica chunk ID again which already has a local directory, shouldn't it just clobber it?

Reply:
I was not very happy with this on previous PR. You can use chunk Id for directory, but I suspected that you wanted to avoid too many directories at one level (though you will probably clean them up anyways). In retries I think that if ou already uploaded things then there is no need to re-upload them again?

Reply:
I've changed this to use `last_update_time` now.

If we later decide that all of the replica chunks should be stored in a flat structure by their ID, we can add this later as an option to the class. In the case where the files are being kept as long-term backups in-place, then the dates should probably be kept to avoid excessive numbers of directories at a single level.